### PR TITLE
Render to FBO.

### PR DESCRIPTION
**Pull request by panzergame**
This branch is to always render to FBOs to keep more control over the FBO quality/size and reuse some of FBO attachment in shaders.

To allow render to FBO some functions was added in the rasterizer to Bind/Unbind/Blit and get textures of an FBO.

There's many different FBO to ping pong for filters, stereo and multisample resolve. They are:
- RAS_OFFSCREEN_RENDER:
   - The first render target bound, it can use multisample.
- RAS_OFFSCREEN_FILTER0
    - One of the filter FBO used for ping pong, it doesn't use multisample. Used for the first blit from RAS_OFFSCREEN_RENDER in case of multisample.
- RAS_OFFSCREEN_FILTER1
    - One of the filter FBO used for ping pong, it doesn't use multisample as RAS_OFFSCREEN_FILTER0.
- RAS_OFFSCREEN_EYE_LEFT0
     - The first render target bound for the stereo left eye, only in case of anaglyph or interlaced stereo. It can use multisample.
- RAS_OFFSCREEN_EYE_RIGHT0
     - The final target of stereo left eye, only in case of anaglyph or interlaced stereo. It doesn't use multisample.
- RAS_OFFSCREEN_EYE_LEFT1
     - The first render target bound for the stereo right eye, only in case of anaglyph or interlaced stereo. It can use multisample.
- RAS_OFFSCREEN_EYE_RIGHT1
     - The final target of stereo right eye, only in case of anaglyph or interlaced stereo. It doesn't use multisample.
- RAS_OFFSCREEN_FINAL
    - The final render target, it doesn't use multisample. Used as output of filters of the last rendered scene.

The major issues was around the filter process. This process behave differently for multisample or not.
With multisamples:
```
n = 0
blit RAS_OFFSCREEN_RENDER to RAS_OFFSCREEN_FILTER0
for filter:
    bind RAS_OFFSCREEN_FILTERn color texture
    bind RAS_OFFSCREEN_FILTERn depth texture
    if final filter:
        bind RAS_OFFSCREEN_RENDER
    else
        bind RAS_OFFSCREEN_FILTER(1-n)
    render shader
    n = 1 - n
```
Without multisample
```
n = 0
for filters:
    if first filter:
        bind RAS_OFFSCREEN_RENDER color texture
        bind RAS_OFFSCREEN_RENDER depth texture
    else:
        bind RAS_OFFSCREEN_FILTERn color texture
        bind RAS_OFFSCREEN_FILTERn depth texture

    if final filter:
        bind RAS_OFFSCREEN_RENDER
    else
        bind RAS_OFFSCREEN_FILTER(1-n)

    render shader
    n = 1 - n
```

For stereo mode anaglyph and interlaced. Two render are made, one per eye. During these render RAS_OFFSCREEN_RENDER is replace by RAS_OFFSCREEN_EYE_LEFT0 or RAS_OFFSCREEN_EYE_RIGHT0, and RAS_OFFSCREEN_FINAL by RAS_OFFSCREEN_EYE_LEFT1 or RAS_OFFSCREEN_EYE_RIGHT1.
At the end of the render of each eyes the filters are proceed. When all the scene and eyes are rendered a quad with shader for the used stereo mode is draw to the screen.

### CODE REVIEW

**Comment by youle31** [comment]
I have a crash in docks level at ge exit that doesn't happen in master (only on ge_fbo)

**Comment by panzergame** [comment]
@youle31 : The crash should be fixed now. A bug about VBO was fixed in master but ge_fbo wasn't yet rebased upon.

**Comment by youle31** [comment]
For me all is working fine :) Congrats and many thanks for your awesome job!!

**Comment by panzergame** [comment]
There's no memory leak if i test with valgrind and a lot of different files.

**Comment by panzergame** [comment]
@youle31, @DCubix, @lordloki : I would like merge this branch tomorrow.

File: source/gameengine/Rasterizer/RAS_OpenGLRasterizer/RAS_OpenGLRasterizer.cpp [hunk 1]
```diff
@@ -208,6 +210,181 @@ void RAS_OpenGLRasterizer::ScreenPlane::Render()
...
+							   index == RAS_OFFSCREEN_EYE_LEFT0 ||
+							   index == RAS_OFFSCREEN_EYE_RIGHT0;
+
+		// Get FBO mode : render buffer support for multisampled FBOs.
+		int mode = GPU_OFFSCREEN_MODE_NONE;
+		if (sampleofs && (m_samples > 0)) {
+			mode = GPU_OFFSCREEN_RENDERBUFFER_COLOR | GPU_OFFSCREEN_RENDERBUFFER_DEPTH;
+		}
+
+		m_offScreens[index] = GPU_offscreen_create(m_width, m_height, sampleofs ? m_samples : 0, (GPUHDRType)m_hdr, mode, NULL);
```
Comment (panzergame) [inline source/gameengine/Rasterizer/RAS_OpenGLRasterizer/RAS_OpenGLRasterizer.cpp]: Some hardware doesn't support high MSAA settings combined with HDR settings. Should we make a loop downgrading these settings while there's an error ?

**Comment by youle31** [comment]
good idea

File: source/blender/gpu/intern/gpu_framebuffer.c [hunk 2]
```diff
@@ -312,6 +313,18 @@ void GPU_framebuffer_bind_no_save(GPUFrameBuffer *fb, int slot)
 	GG.currentfb = fb->object;
 }
 
+void GPU_framebuffer_bind_simple(GPUFrameBuffer *fb)
+{
+	if (GG.currentfb != fb->object) {
+		glBindFramebufferEXT(GL_FRAMEBUFFER_EXT, fb->object);
+		/* last bound prevails here, better allow explicit control here too */
```
Comment (lordloki) [inline source/blender/gpu/intern/gpu_framebuffer.c]: better to allow explicit....

File: source/blender/gpu/intern/gpu_framebuffer.c [hunk 3]
```diff
@@ -444,15 +488,126 @@ void GPU_framebuffer_blur(
...
+
+	glGenRenderbuffers(1, &rb->bindcode);
+
+	if (!rb->bindcode) {
+		if (err_out) {
+			BLI_snprintf(err_out, 256, "GPURenderBuffer: render buffer create failed: %d",
+				(int)glGetError());
+		}
+		else {
+			fprintf(stderr, "GPURenderBuffer: render buffer create failed: %d\n",
```
Comment (lordloki) [inline source/blender/gpu/intern/gpu_framebuffer.c]: render buffer creation failed

File: source/blender/gpu/intern/gpu_framebuffer.c [hunk 4]
```diff
@@ -466,41 +621,78 @@ GPUOffScreen *GPU_offscreen_create(int width, int height, int samples, char err_
...
+			 * when it's not supported */
+		    (!GLEW_ARB_texture_multisample && (!(mode & GPU_OFFSCREEN_RENDERBUFFER_COLOR) || !(mode & GPU_OFFSCREEN_RENDERBUFFER_DEPTH))) ||
 		    /* Only needed for GPU_offscreen_read_pixels.
 		     * We could add an arg if we intend to use multi-sample
 		     * offscreen buffers w/o reading their pixels */
-		    !GLEW_EXT_framebuffer_blit ||
+		    !GLEW_EXT_framebuffer_blit
+
+	/* Some GPUs works even without this extension. */
+#if 0
```
Comment (lordloki) [inline source/blender/gpu/intern/gpu_framebuffer.c]: if it is disabled, shoud it be deleted?

Comment (panzergame) [inline source/blender/gpu/intern/gpu_framebuffer.c]: It's left only for futur issues related to this.

**Comment by lordloki** [comment]
Seems Ok to me. Very good work. +1

**Comment by panzergame** [comment]
The pull request is closed but the ge_fbo branch is still here for stereo bug fix.

**Pull request closed without merging**